File: python/crypto/actor_keys.py
"""
Actor key pair management for Mind Protocol.

X25519 key generation, storage, and loading using PyNaCl.
Key files use the same .b64 format as the JS implementation
(base64-encoded raw 32-byte keys).

Co-Authored-By: Tomaso Nervo (@nervo) <nervo@mindprotocol.ai>
"""

import os
import base64

from nacl.public import PrivateKey


def generate_actor_key_pair() -> dict:
    """Generate an X25519 key pair for an actor.

    Returns:
        dict with keys:
            - "public_key": 32-byte public key (bytes)
            - "private_key": 32-byte private key (bytes)
    """
    private_key = PrivateKey.generate()
    public_key = private_key.public_key

    return {
        "public_key": bytes(public_key),
        "private_key": bytes(private_key),
    }


def store_actor_keys(keys_dir: str, key_pair: dict) -> None:
    """Write an actor's key pair to .b64 files on disk.

    Creates the directory if it does not exist. Writes two files:
        - <keys_dir>/public_key.b64   (base64 of 32 raw bytes)
        - <keys_dir>/private_key.b64  (base64 of 32 raw bytes)

    Args:
        keys_dir: Directory path where key files will be written.
        key_pair: dict with "public_key" and "private_key" as bytes.

    Raises:
        ValueError: If the key pair is missing required fields.
    """
    if "public_key" not in key_pair or "private_key" not in key_pair:
        raise ValueError("key_pair must contain 'public_key' and 'private_key'")

    os.makedirs(keys_dir, exist_ok=True)

    pub_b64 = base64.b64encode(key_pair["public_key"]).decode("ascii")
    priv_b64 = base64.b64encode(key_pair["private_key"]).decode("ascii")

    pub_path = os.path.join(keys_dir, "public_key.b64")
    priv_path = os.path.join(keys_dir, "private_key.b64")

    with open(pub_path, "w") as f:
        f.write(pub_b64)

    with open(priv_path, "w") as f:
        f.write(priv_b64)


def load_actor_keys(keys_dir: str) -> dict:
    """Load an actor's key pair from .b64 files on disk.

    Reads:
        - <keys_dir>/public_key.b64
        - <keys_dir>/private_key.b64

    Args:
        keys_dir: Directory containing the key files.

    Returns:
        dict with keys:
            - "public_key": 32-byte public key (bytes)
            - "private_key": 32-byte private key (bytes)

    Raises:
        FileNotFoundError: If either key file is missing.
        ValueError: If the decoded key lengths are incorrect.
    """
    pub_path = os.path.join(keys_dir, "public_key.b64")
    priv_path = os.path.join(keys_dir, "private_key.b64")

    with open(pub_path, "r") as f:
        pub_bytes = base64.b64decode(f.read().strip())

    with open(priv_path, "r") as f:
        priv_bytes = base64.b64decode(f.read().strip())

    if len(pub_bytes) != 32:
        raise ValueError(f"Public key must be 32 bytes, got {len(pub_bytes)}")
    if len(priv_bytes) != 32:
        raise ValueError(f"Private key must be 32 bytes, got {len(priv_bytes)}")

    return {
        "public_key": pub_bytes,
        "private_key": priv_bytes,
    }

File: lib/crypto/actor_keys.js
/**
 * Actor Keys — X25519 key pair generation and storage for Mind Protocol actors.
 *
 * Uses libsodium sealed-box-compatible key pairs (crypto_box_keypair).
 * Keys are stored as base64-encoded 32-byte files: public_key.b64 / private_key.b64.
 *
 * Co-Authored-By: Tomaso Nervo (@nervo) <nervo@mindprotocol.ai>
 */

'use strict';

const fs = require('fs');
const path = require('path');
const sodium = require('libsodium-wrappers');

// --- Constants ---

const PUBLIC_KEY_FILE = 'public_key.b64';
const PRIVATE_KEY_FILE = 'private_key.b64';
const KEY_LENGTH = 32;

// --- Functions ---

/**
 * Generate an X25519 key pair suitable for sealed-box encryption.
 *
 * Must be called after `await sodium.ready`.
 *
 * @returns {Promise<{ publicKey: Buffer, privateKey: Buffer }>} 32-byte key pair
 */
async function generateActorKeyPair() {
  await sodium.ready;

  const kp = sodium.crypto_box_keypair();

  return {
    publicKey: Buffer.from(kp.publicKey),
    privateKey: Buffer.from(kp.privateKey),
  };
}

/**
 * Store an actor key pair to disk as base64-encoded .b64 files.
 *
 * Creates the directory if it does not exist.
 *
 * @param {string} keysDir - Directory to write key files into
 * @param {{ publicKey: Buffer, privateKey: Buffer }} keyPair
 * @returns {Promise<void>}
 */
async function storeActorKeys(keysDir, keyPair) {
  if (!keyPair || !Buffer.isBuffer(keyPair.publicKey) || !Buffer.isBuffer(keyPair.privateKey)) {
    throw new Error('keyPair must contain publicKey and privateKey Buffers');
  }
  if (keyPair.publicKey.length !== KEY_LENGTH) {
    throw new Error(`publicKey must be ${KEY_LENGTH} bytes, got ${keyPair.publicKey.length}`);
  }
  if (keyPair.privateKey.length !== KEY_LENGTH) {
    throw new Error(`privateKey must be ${KEY_LENGTH} bytes, got ${keyPair.privateKey.length}`);
  }

  fs.mkdirSync(keysDir, { recursive: true });

  const pubPath = path.join(keysDir, PUBLIC_KEY_FILE);
  const privPath = path.join(keysDir, PRIVATE_KEY_FILE);

  fs.writeFileSync(pubPath, keyPair.publicKey.toString('base64'), 'utf8');
  fs.writeFileSync(privPath, keyPair.privateKey.toString('base64'), 'utf8');
}

/**
 * Load an actor key pair from base64-encoded .b64 files on disk.
 *
 * @param {string} keysDir - Directory containing public_key.b64 and private_key.b64
 * @returns {Promise<{ publicKey: Buffer, privateKey: Buffer }>} 32-byte key pair
 */
async function loadActorKeys(keysDir) {
  const pubPath = path.join(keysDir, PUBLIC_KEY_FILE);
  const privPath = path.join(keysDir, PRIVATE_KEY_FILE);

  if (!fs.existsSync(pubPath)) {
    throw new Error(`Public key file not found: ${pubPath}`);
  }
  if (!fs.existsSync(privPath)) {
    throw new Error(`Private key file not found: ${privPath}`);
  }

  const pubB64 = fs.readFileSync(pubPath, 'utf8').trim();
  const privB64 = fs.readFileSync(privPath, 'utf8').trim();

  const publicKey = Buffer.from(pubB64, 'base64');
  const privateKey = Buffer.from(privB64, 'base64');

  if (publicKey.length !== KEY_LENGTH) {
    throw new Error(`Invalid public key length: expected ${KEY_LENGTH}, got ${publicKey.length}`);
  }
  if (privateKey.length !== KEY_LENGTH) {
    throw new Error(`Invalid private key length: expected ${KEY_LENGTH}, got ${privateKey.length}`);
  }

  return { publicKey, privateKey };
}

// --- Exports ---

module.exports = {
  PUBLIC_KEY_FILE,
  PRIVATE_KEY_FILE,
  KEY_LENGTH,
  generateActorKeyPair,
  storeActorKeys,
  loadActorKeys,
};
